### DISCUSSION TITLE
Zustand store doesn't work when putting to external package

### DISCUSSION THREAD

**Discussion by roman-rookout** [pmndrs/zustand#2870] [Q&A] [answered]
Hi everyone. We have a set of different webViews, they are all independent react apps that can have a shared codebase, but run in an isolated manner. We started to build a lerna monorepo repository and split the codebase into different reusable packages.
We have multiple reusable infra zustand stores that will be placed in the shared package and be imported from different applications. Here, we encounter issues. After moving the store into the package, the app stopped loading. I built a test repo to reproduce the issue from the clean repository.

I have a package with a store:
`const useTestStore = create<Store>((set) => ({`
`      count: 0,`
`    increment: () => set((state) => ({ count: state.count + 1 })),`
`    decrement: () => set((state) => ({ count: state.count - 1 })),`
`}));`

I'm using it in the main component then 
`function App() {`
 `     const { count, increment, decrement } = useTestStore();`
     ...
`       <p>{count}</p>`
`      <button onClick={() => increment()}>`
`        Increment`
`      </button>`

What I see is  a warning about InvalidHooks usage outside of react component which is weird and  "TypeError: Cannot read properties of null (reading 'useRef')".
It points to the row - const { count, increment, decrement } = useTestStore();

![Screenshot 2024-11-26 at 18 04 56](https://github.com/user-attachments/assets/ea6a8e2d-557a-48e8-9594-44073690e178)

So any usage of hook will fail also useTestStore((state) => state.count)
Only "non hook" method like `useTestStore.getState().count` will work.
Would appreciate for any input. Maybe we should reconsider the way we bundle the library, we are using Vite.

The PoC repo is here: 
https://github.com/roman-rookout/zustand-package-test

**Comment by dai-shi** [comment]
> You might have more than one copy of React in the same app

It sounds like it's that case.

**Reply by roman-rookout** [reply]
That is why I tried to isolate the issue in a clean repo to reproduce. I have peer dependency only in the library/package and only one reference to react and zustand from the application itself. I've been playing with bundling for ages, to make sure I don't have duplicate version of packages. Also here is the output of npm list react which shows I am not referencing anything excessive
![image](https://github.com/user-attachments/assets/cb3f29de-a179-4de1-bcb0-f387b509c9d2)

**Reply by roman-rookout** [reply]
I couldn't find on the internet much useful information about usage zustand in monorepo in a shared package, seemed to me like it would've been a pretty common solution

**Reply by roman-rookout** [reply]
I updated vite.config with additional resolvers. Weirdly it changed the error to `Uncaught TypeError: Cannot read properties of null (reading 'useSyncExternalStore')`
![image](https://github.com/user-attachments/assets/991a74b9-2a13-4f76-837c-3573680ffcb3)

The thing is, it is a very clean repository with minimum things that could've caused a wrong reaction, and I don't see any duplication in the node_modules folder or npm list.

**Reply by roman-rookout** [reply]
Also tried to install` use-sync-external-store `, didn't help even after playing with adding alias in vite

**Reply by Trung-Pham-Modec** [reply]
I have the same issue, @roman-rookout Did you find the fix for this issue?

**Reply by roman-rookout** [reply, marked as answer]
@Trung-Pham-Modec 
in the end, it was connected to react duplication issue and my problem was - I misplaced rollupOptions in vite config at my lib's config. The correct vite config for the library is:

![image](https://github.com/user-attachments/assets/13b3474a-b8f5-4b31-b0e4-e698682d14e5)

**Comment by Temetz** [comment]
This may help someone in the future...

If you are doing manual chunking, zustand must be included in the same chunk as react and react-dom.

```
export default defineConfig({
  build: {
    rollupOptions: {
      output: {
        compact: true,
        manualChunks: {
          sentry: ["@sentry/browser"],
          // CHECK THIS!
          react: ["react", "react-dom", "zustand"],
          mui: ["@mui/material", "@emotion/react", "@emotion/styled"],
        },
      },
      treeshake: "recommended",
    },
  },
```
